File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char> mp;
        mp['(']= ')'; mp['{']= '}'; mp['[']= ']';
        stack<char> stk;
        for(char ch: s){
            if(mp.find(ch)==mp.end()){
                if(!stk.empty() && mp[stk.top()]==ch){
                    stk.pop();
                    continue;
                }
                else{
                    return false;
                }
            }
            else{
                stk.push(ch);
            }
        }
        return stk.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stk = new Stack<>();
        Map<Character,Character> dict = new HashMap<>();
        dict.put('(',')'); dict.put('{','}'); dict.put('[',']');
        char ch;
        for(int i=0;i<s.length();i++){
            ch = s.charAt(i);
            if(ch=='(' || ch=='{' || ch=='['){
                stk.push(ch);
            }
            else{
                if(stk.isEmpty() || dict.get(stk.peek())!=ch){
                    return false;
                }
                stk.pop();
            }
        }
        return stk.isEmpty();
    }
}